### DISCUSSION TITLE
signature validation error

### DISCUSSION THREAD

**Discussion by Metal-Bat** [MatthiasValvekens/pyHanko#317] [Q&A] [answered]
hi every one

i am facing an issue and needed help
i am using 'pyhanko sign validate --pretty-print --no-diff-analysis --no-strict-syntax download.pdf'

and i get  these two error:

===================
Field 1: Signature1
===================
An error occurred while validating this signature: The algorithm sha256_rsa is not allowed by the current usage policy. Reason: Key size 1024 for algorithm rsassa_pkcs1v15 is considered too small; policy mandates >= 2048. [AdESIndeterminate.CRYPTO_CONSTRAINTS_FAILURE]
2023-09-11 16:42:03,330 - pyhanko.sign.validation.generic_cms - WARNING - Failed to build path

===================
Field 3: Signature2
===================
An error occurred while validating this signature: The algorithm sha1 is not allowed by the current usage policy. [AdESIndeterminate.CRYPTO_CONSTRAINTS_FAILURE]
Error: Validation failed

is this some kind of security check?
and what is the solution for two error above 

your help is really appreciated, Thank you

**Comment by Metal-Bat** [comment]
i managed to read docs and 
i kinda have this code

***

import io
from base64 import b64decode

from pyhanko.pdf_utils.reader import PdfFileReader
from pyhanko.sign.validation import validate_pdf_signature
from pyhanko_certvalidator import ValidationContext

bytes = io.BytesIO(b64decode(base_64_data, validate=True))

pdf_file = PdfFileReader(bytes)
validatino = ValidationContext()
for sig in pdf_file.embedded_signatures:
    print("*" * 10, sig, "*" * 10)
    try:
        status = validate_pdf_signature(sig)
        print(status.pretty_print_details())

    except Exception as e:
        print(e)

***

i know no one use base64 datas these days and i kinda get i need to write my own policy for validation but o have strugle with this and kinda feels the policy is there and i just need to find it , b.c i use pdfsig in my terminal and everything works fine

also your help is really appreciated

**Reply by MatthiasValvekens** [reply, marked as answer]
I struggle a bit to see where the link with base64 is, but OK :) 

Anyway, the thing you have to override here is the algorithm usage policy. That's a property of the `ValidationContext`.

The default one rejects RSA with small keys and SHA-1, as you noticed. Those aren't sound choices in this day and age... but if you absolutely must, you can use this policy to effectively disable all such checks: https://pyhanko.readthedocs.io/en/latest/api-docs/pyhanko_certvalidator.html#pyhanko_certvalidator.policy_decl.AcceptAllAlgorithms

**Comment by Metal-Bat** [comment]
You are Amazing
thank you
one more
help and i promisse i won't be a headech
i know it seems i work on somethig super wierd and honstly i am

when i validate pdf 
```

status = validate_pdf_signature(
            embedded_sig=sig,
            ts_validation_context=validation,
            signer_validation_context=validation,
            skip_diff=True,
        )

```

it just printed out no biggie
`pyhanko_certvalidator.errors.PathBuildingError:  Unable to build a validation path for the certificate`

and more thing in pretty print it shows this
```

Bottom line
-----------
The signature is judged INVALID.

```
and in
`print(status)`
this comes up
`PdfSignatureStatus(intact=True, valid=True, trust_problem_indic=<AdESIndeterminate.NO_CERTIFICATE_CHAIN_FOUND: 11>
`
how can i validate it is now kind of wierd
(it has oscp in it)

i mean this is the out put of

`pdfsig download\ \(2\).pdf
`
```
Digital Signature Info of: download (2).pdf
Signature #1:
  - Signature Field Name: Signature1
  - Signer Certificate Common Name: # some fake data
  - Signer full Distinguished Name: # some fake data
  - Signing Time: Dec 26 2022 21:06:47
  - Signing Hash Algorithm: SHA-256
  - Signature Type: adbe.pkcs7.detached
  - Signed Ranges: [0 - 41698], [79588 - 91452]
  - Not total document signed
  - Signature Validation: Signature is Valid.
  - Certificate Validation: Unknown issue with Certificate or corrupted data.
```

maybe i just got confused between Signature Validation and  Certificate Validation

i hope i didn't bother you and thanks
